networkx/linalg/algebraicconnectivity.py

```diff
@@ -244,7 +244,7 @@ def P(x):
             W -= (W.T * X * X.T).T
             project(W)
             # Compute the diagonal of P * L * P as a Jacobi preconditioner.
-            D = L.diagonal()
+            D = L.diagonal().astype(float)
             D += 2. * (asarray(X) * asarray(W)).sum(axis=1)
```

Comment:
My bad for making the code really obscure. `D` is used in line 205.

The code is written in such a way that `solver` is initialized only once. This is necessary for the Cholesky and LU solvers as their factorization steps are expensive. But this creates a problem for the PCG solver as its preconditioner changes over time. To hack around this problem, the preconditioner is supplied as a lambda that references `D`. By changing `D`, the preconditioner is also changed.

A more readable design is to reinitialize the PCG solver in each TraceMIN iteration supplying `D` explicitly. This should avoid people recognizing `D` as unused variable by accident.